### PR TITLE
Added export-values handling to dependencies

### PR DESCRIPTION
**Pull request by withnale**
NB: This is a reworked implementation of the functionality found in PR #3243 

---

See issue #3242 

There is a need for an `export-values` method when working with requirements.

The attached PR approaches this by mimicking the `import-values` functionality but in the other direction - allowing you to specify the tree in the parent chart you wish to export into the values of the child chart. It also allows you to _ringfence_ settings to a duplicate (aliased) subchart, as can be seen in the example below:

```
dependencies:
      - name: subchart1
        repository: http://localhost:10191
        version: 0.1.0
        tags:
          - front-end
          - subchart1
        export-values:
          - child: release
            parent: release.old

      - name: subchart1
        repository: http://localhost:10191
        version: 0.1.0
        export-values:
          - child: release
            parent: release.new
        alias: subchart1a
```
Obviously, this simple use case can be achieved through other methods - it's when there are many requirements using a mix & match of various parent values that this change become effective.

NB: We should really rename these methods but not sure what would be a good plan. Should it go from `processDependencyImportValues` to `processDependencyImportExportValues`, something else, or should we just leave them with their existing names?

### CODE REVIEW

**Comment by withnale** [comment]
Can I get a review?

**Review by bacongobbler — APPROVED** [review]

**Comment by bacongobbler** [comment]
Code LGTM. I'm going to hold off on merging until another maintainer reviews this.

In the meantime, would you mind writing some documentation that covers this new function? See https://github.com/helm/helm-www#how-to-edit-the-helm-docs

Let me know if you have any questions

**Comment by withnale** [comment]
Do you want me to change the naming scheme for the functions like `processImportValues` (since they now process the exportValues as well) and if you, to what? Or are you happy with this just being a bit of _history_ in the project.

Will add some docs.

File: pkg/chartutil/dependencies.go
Comment (hickeyma) [inline pkg/chartutil/dependencies.go]: Some modification here as follows:
- Move this comment to above : `for _, riv := range r.ExportValues {`
- Add a comment here similar to:  `// handle export and import values per dependency`

**Review by hickeyma — CHANGES_REQUESTED** [review]
LGTM, thanks for you effort and persistence @withnale [great handle btw, esp. if its based off that movie!:)]. I have a small nit inline.

Wrt https://github.com/helm/helm/pull/7477#issuecomment-589573764 about separating out into distinct function, I think this is a good idea for maintenance later.  Suggestion of new funcs:
- `processExportValues`
- `processDependencyExportValues`

**Comment by bacongobbler** [comment]
> Will add some docs.

Bump. Can you please provide some documentation here? That way we can give this PR a proper round of testing.

As much as I like to think I understand everything that goes on in Helm-land, some instructions on how to test this new feature would be greatly appreciated. I haven't manually tested this yet. I've only given this a code review. :)

FYI, we're planning to cut a release candidate of Helm 3.2.0 a week prior to KubeCon EU. If you have time to do that soon, we can expedite the review and see if we can get this in for 3.2.0. Otherwise we may have to punt this over to the next release cycle.

Thanks!

**Comment by hickeyma** [comment]
@withnale As per @bacongobbler https://github.com/helm/helm/pull/7477#issuecomment-593596776, can you provide doc on usage so we can progress this PR?

**Comment by anlutro** [comment]
I spent a whole day trying to figure out how to achieve exactly this and finally ended up finding this PR. This is an amazing feature that I think is sorely needed. Can I help in some way? If not, this is just a shameless ping.

**Comment by hickeyma** [comment]
Ping @withnale

**Comment by withnale** [comment]
Sorry. Been a bit preoccupied over the last couple of weeks. Finally got to look at this change again.

The PR has changed a little. @hickeyma contacted me directly as stated that it was having trouble testing it and I noticed that although the unit tests I'd written passed, it didn't work when running helm template. I also noticed that the original PR only worked for references that were Table and not for strings. 

> LGTM, thanks for you effort and persistence @withnale [great handle btw, esp. if its based off that movie!:)]. I have a small nit inline.

> Wrt [#7477 (comment)](https://github.com/helm/helm/pull/7477#issuecomment-589573764) about separating out into distinct function, I think this is a good idea for maintenance later. Suggestion of new funcs:
> 
> * `processExportValues`
> * `processDependencyExportValues`

I was thinking that at present this logic actually fits relatively nicely into one function but the function names don't now reflect what it is doing. How about instead `processPassingValues` and `processDependencyPassingValues`?

I've pushed the PR as another WIP to see if anyone can suggest a nicer way of handling the value case, other than using `LastIndex` the way I have done.

**Comment by bacongobbler** [comment]
Bump.

> Bump. Can you please provide some documentation here? That way we can give this PR a proper round of testing.

Without proper documentation, we cannot test this the way you intended it to work.

**Comment by withnale** [comment]
I've taken an initial stab at documenting this at https://github.com/helm/helm-www/pull/680

I'll rebase to fix the conflicts too.

**Comment by hickeyma** [comment]
@withnale Thanks for resolving conflicts and providing documentation.

I have been testing it manually but it does not seem to be working as I expect.

I have tested using a simple parent chart (`chrt-7477`) and subchart (`subchrt1`) as follows:

```console
$ cat chrt-7477/values.yaml 
release: "Parent release"

$ cat chrt-7477/Chart.yaml 
apiVersion: v2
name: chrt-7477
description: A Helm chart for Kubernetes
[...]

dependencies:
- name: subchrt1
  repository: file://../subchrt1
  version: 0.1.0
  export-values:
    - parent: release
      child: release

$ cat chrt-7477/templates/configmap.yaml 
apiVersion: v1
kind: ConfigMap
metadata:
  name: {{ .Release.Name }}-configmap
data:
  myvalue: "Hello World"
  rel: {{ .Values.release }}

$ cat subchrt1/values.yaml 
release: "subchrt1 release" 

$ cat subchrt1/Chart.yaml 
apiVersion: v2
name: subchrt1
description: A Helm chart for Kubernetes
[...]

appVersion: 1.16.0

$ cat subchrt1/templates/configmap.yaml 
apiVersion: v1
kind: ConfigMap
metadata:
  name: {{ .Release.Name }}-configmap-subchrt1
data:
  myvalue: "Hello World from subchrt1"
  rel: {{ .Values.release }}
```

I would expect the output of `subchrt1` data property `rel` to be `Parent release`:

```console
$ helm3 dep update chrt-7477
Hang tight while we grab the latest from your chart repositories...
...Successfully got an update from the "stable" chart repository
Update Complete. ⎈Happy Helming!⎈
Saving 1 charts
Deleting outdated charts

$ helm3 install chrt-7477 chrt-7477/ --dry-run
NAME: chrt-7477
LAST DEPLOYED: Fri Jun 19 17:21:49 2020
NAMESPACE: default
STATUS: pending-install
REVISION: 1
TEST SUITE: None
HOOKS:
MANIFEST:
---
# Source: chrt-7477/charts/subchrt1/templates/configmap.yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: chrt-7477-configmap-subchrt1
data:
  myvalue: "Hello World from subchrt1"
  rel: subchrt1 release
---
# Source: chrt-7477/templates/configmap.yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: chrt-7477-configmap
data:
  myvalue: "Hello World"
  rel: Parent release
```

**Review by hickeyma — CHANGES_REQUESTED** [review]
@withnale I have tested the PR manually and it doesn't seem to be working as expected. Refer to https://github.com/helm/helm/pull/7477#issuecomment-646780137 for more details.

**Comment by hickeyma** [comment]
@withnale We are aiming to cut 3.3.0 release shortly, so it would be great if you could address the feedback when you can. Thanks.

**Comment by bacongobbler** [comment]
3.3.0-rc.1 is being cut tomorrow, so we're removing this from the milestone.

**Comment by bacongobbler** [comment]
any update here?

**Comment by NMFR** [comment]
This would be great. Anything i can do to help with this?

**Comment by emilchitas** [comment]
Are there any new updates on this. 
It would be a much appreciated feature.

**Comment by dejan9393** [comment]
I'm highly anticipating this feature, as are many others, I believe. I find myself very often being constrained by the same issue of attempting to re-use parent chart values in my subcharts, leading to me spending countless hours scouring the net looking for workarounds/possible new solutions to this issue.

I've been faced with this multiple times over the past 3 years, and recently came across the import-values feature, which is great! I think this would be an excellent complement to that feature!

**Comment by bacongobbler** [comment]
The last code review conducted was exactly 13 months ago. Changes were requested. No changes have been made since. I think it’s safe to say it’s unlikely the OP will be prepared to make changes any time soon.

If this is something you’d like to see, I would suggest taking on the work that was applied earlier and re-submit a new PR.

For now, I’m going ahead and closing this as stale.

**Pull request closed without merging**
